### ISSUE TITLE
Rework map area-setting sub-UI needs to be more more preset-focussed

### ISSUE THREAD

**Issue by mvl22** [cyclestreets/cyclescape#661]
A wholescale rethinking of the map area-setting sub-UI is planned.

Currently, setting areas, which is critical for getting an effective user account (as well as setting issues correctly) currently requires proficiency with maps, which is not a universal skill.

Map drawing (as at present) should remain a first-class option in this sub-UI but it should not be the default mode of setting a geographical area.

Instead, the proposal is to get key information from the user (e.g. your home and work postcodes/streetnames), and the UI will then suggest useful likely desired locations, that they can easily 'tick', resulting in the shape appearing on the map:

* 'Near where I live' (something like the 500m circle around home)
* 'My ward' (e.g. Romsey Ward in Cambridge)
* 'My route to work' (obviously using the CycleStreets API, result draggable to fix)
* 'My city' (e.g. Cambridge; though users should be warned if this is likely to give them large amounts of activity)
* 'My group's area' (again, users should be warned if this is likely to give them large amounts of activity)
* Search for a name (e.g. if they search for a road name or part of a city, it would put that on the map; note that the CycleStreets Geocoder API now can return full boundaries, so this should now be easy to achieve)
* Plus as now: 'Draw manually on a map'
* etc.

Essentially this should enable users to get going quickly, with only a few clicks, giving them a sensible set of defaults for where they live and travel to, and with the user being clear about the likely amount of activity this would see them involved in (so they can be reassured about not being overwhelmed).

The above will probably require a mix of: MySociety's MapIt API, the CycleStreets Geocoder V2 API, and the CycleStreets routing API.

Now that the underlying changes to replace OpenLayers2 with Leaflet are in place, we can start moving forward on this.

**Comment by jf1** [comment]
Just ask if you'd like to offer more feedback on this or clarify any of the below, but for now...

> (e.g. your home and work postcodes/streetnames)

Could this include eg 'transport interchange' for people who are more interested in the cycle from station to work, and perhaps never cycle all the way from home to work. Also, something like 'regular destination' would be more inclusive than 'work'.

> 'My ward' (e.g. Romsey Ward in Cambridge)

I hope there'll be multi-select options, including across LA boundaries. Perhaps enabling users to merge two feature collections is the best way to enable users to add multiple wards from more than one LA into a single Cyclescape area. It'd also be useful to multi-select from a map, as people know where they ride but don't necessarily know the ward names they're riding through.

> 'My route to work'

I hope this will include an eg 50m strip either side of the Cyclestreets route, to ensure slightly misplaced issues and works that are close enough to have an effect are both included

> 'My city'

I'd be interested to hear how you plan to deal with Manchester, or Greater Manchester, or the bit of the City of Salford that even they describe as being within Manchester city-centre. Not sure there's a way to do anything better than: My CA/County, My LA, My ward(s), etc.
If practical it'd be good to include Westminster constituencies too.

> 'My group's area' (again, users should be warned if this is likely to give them large amounts of activity)

This is where I'd like to make a new suggestion that could solve #665 too.  New feature request is in #666 "Enable per-area and per-group opt in/out of email notifications and planning applications".

> Search for a name... API now can return full boundaries

For reference the [UGC boundaries](http://geoportal.statistics.gov.uk/datasets?q=ultra+generalised+clipped) would suffice for me (although for wards SGC may be the simplest version available). Might this benefit from a radio-select for: Default; Include administrative boundaries; Administrative boundaries only? Manchester's probably a good test case again - people use it to refer to: the point at the centre of the city; the city-centre area; the City of Manchester LA; and the Greater Manchester CA.

**Comment by mvl22** [comment]
> I hope this will include an eg 50m strip either side of the Cyclestreets route

Already happens. Draw a route in your location settings and you'll see it actually becomes an area.

**Comment by jf1** [comment]
> Draw a route in your location settings and you'll see it actually becomes an area.

Thanks, I don't remember it doing that previously, although I'm not really a user of this feature. Do people still have to draw routes manually using straight lines or can we now do start and end points with OSM routing?

**Comment by nikolai-b** [comment]
All these depend on having a postcode so do we change https://www.cyclescape.org/settings/locations to have have a "post code" input (or even ask [the browser](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/geolocation) for the users location).  So entering the postcode (or allowing the browser to geolocate) would stop some tickboxes being disabled (i.e. when we have my regular route it would require a start and end post code or that checkbox would be disabled).

Can we start with doing just 2 of these please?

* 'Near where I live' (something like the 500m circle around home)
* 'My ward' (e.g. Romsey Ward in Cambridge)

Presumably as you want a tick option then this would be a multi-location geometry (e.g. potentially a route and an area).  When one is ticked the shape would appear on the map, with an internal id so that it could also be removed from the map if un-ticked.  It could also be edited.  The "Near where I live" would be a case of adding a few points (8?) to make an octagon around the persons home.

Also @jf1 the additions you mention are sensible regarding merging ward geometries but they come at the expense of limited time and resources.  I would be very pleased to use the UGC [ward boundaries](http://geoportal.statistics.gov.uk/datasets/9507292be10e43c7abe9d3be36db6c41_4) if @mvl22 is happy as that would mean one import and then we wouldn't need to have a API call.

**Comment by mvl22** [comment]
> Can we start with doing just 2 of these please?

Yes, I definitely see this as something that will build up over time, but that we can get some basic stuff done like this.

> Presumably as you want a tick option then this would be a multi-location geometry (e.g. potentially a route and an area). When one is ticked the shape would appear on the map, with an internal id so that it could also be removed from the map if un-ticked. It could also be edited. The "Near where I live" would be a case of adding a few points (8?) to make an octagon around the persons home.

Yep, all sounds great.

>  I would be very pleased to use the UGC ward boundaries if @mvl22 is happy as that would mean one import and then we wouldn't need to have a API call.

Yes, that's fine.

Bear in mind that it shouldn't become dependent on these, given other countries' use.

**Comment by nikolai-b** [comment]
This has been started in #672.

I'm unsure about what the UI you envisage looks like, is it something like this?

![ui_test](https://cloud.githubusercontent.com/assets/6289830/21552641/e581b78a-cdfa-11e6-91c7-bdcf3091199c.png)

**Comment by nikolai-b** [comment]
I've added a few more things so we have constituencies and route to destination @mvl22

**Comment by mvl22** [comment]
Thanks - this looks like it is coming on nicely.

Let's get more of the functionality in place then I think I'll be in a better position to think more about an optimal UI. I'm finding it quite hard to judge at the moment - hence why I didn't reply on the original work last week (sorry about that!) Very happy to consider whatever you come up with and work from there.

I definitely think that it would be good to split out Council ward and Constituency.

Another useful one would be 'Groups in my area'.

One thing the UI would ideally do is give the user an idea of how many subscriptions this would be likely to give them.

**Comment by nikolai-b** [comment]
There were failing specs, no need for a new regression test, my bad.

I've moved to HTTPS.  I don't want to add more functionality if I'm going down the wrong path.

> I definitely think that it would be good to split out Council ward and Constituency.

I've added wards but I don't want to add an idea of subscriptions till we are happy with the work that I've done.  If we are allowing multiple overlapping areas than that would be hard to implement and at the moment we let people create locations without giving them a number of emails.

**Comment by nikolai-b** [comment]
Have you had a chance to review @mvl22 ?  With constituency, wards, area around me and route to destination I believe this is a sensible start that could be released.  

Can we get this live and then I can add groups and also an idea of the number of issues in an area for the past month.

**Comment by mvl22** [comment]
This is a huge step forward in terms of the feature set for setting locations - nice work. (Great you got routing working too!) However, the current prototype implementation doesn't yet form something that gives a clear and unambiguous action flow to the user. I think this is quite a difficult interface to get completely right, however, it is starting to crystallise in my mind how it could work.

Before we do that, could you make some changes which will be needed anyway, and remove a whole class of usability problems:

1) Remove the category stuff. I don't think this has ever really been helpful, and it certainly makes no sense now that multiple geometries can end up on the map. I would simply remove the form widget, and set the field to be optional in database schema terms, so that people's existing data is not wiped out, and so that we can use this again in future if required.

2) Add migrate-away protection. Currently it is possible for someone to spend quite a bit of time adjusting the map, see that all the geometries are present, then click on the main menu assuming that they have been set. They then lose all the data. So it needs the usual 'Are you sure you want to leave this page' style javascript.

3) The drop-down highlighting is the wrong way round. Currently all end up grey, with the one you want as white. Instead, the selected item should be the highlighted one. Could you also make sure the drop-down is the same width as the input box itself - it should not be dependent on the entries that appear. This will help to create a clear mental modal that the input box and its drop down form a single semantic control.

Could you fix these up firstly, and I'll then propose a second set of changes which will get us further to a releasable UI.

**Comment by nikolai-b** [comment]
> I think this is quite a difficult interface to get completely right; however, it is starting to crystallise in my mind how it could work.

Agreed, in the sense that I think we can do better that this first draft but I thought it is good to push things forward so thank you very much for the feedback.

> 1. Remove the category control at long last. 

Done in 8e56868f

> 2. Add migrate-away protection. 

Done in badf0212

> 3. The search-as-you-type drop-down highlighting is the wrong way round. Currently all end up grey, with the one you want as white.

Do you mean this:
![screenshot from 2017-01-08 20-42-10](https://cloud.githubusercontent.com/assets/6289830/21753343/0556bb2e-d5e3-11e6-9ee8-fcf988600ef3.png)

if so that is done in 32919bf1

> Could you also make sure the drop-down is the same width as the input box itself - it should not be dependent on the entries that appear. This will help to create a clear mental modal that the input box and its drop down form a single semantic control.

I've done (also in 32919bf1) this but I'm less happy about it, it makes it hard to distinguish between entries if only the last characters are different but see what you think.

 > The map is quite dominant in terms of detail under the overlaid geometries, so that currently it's quite hard to see what has been overlaid. 

Done in 72dbf965

Let me know how the second set of changes look.  

I'll start work on adding nearby groups but there are some questions there, is that a check box for each group in your area (or a multi-select box)?

With the estimated number of issues where would you envisage that to go?  Somewhere near save with:
"Number of new issues in your location over the past month"
?

**Comment by nikolai-b** [comment]
A very basic groups preset is now live 100d6bd2

**Comment by mvl22** [comment]
Great!

Here are some UI suggestions - what do you think? I'm not 100% sure myself yet.

What I'm thinking about is a series of boxes on the left, stacked one on top of the other, each of which has a type of geometry addition, using the functionality you've now built. Each box would have the controls relevant to that type of addition. E.g. routing would have start,finish; groups would have list of tickboxes; etc. On the right would be the map, which shows the result. Each box would have some kind of tick icon or arrow icon that essentially 'sends' the type to the map. Or each could just have a checkbox to enable/disable each item.

As almost all the types of location require a start point, at the top there would be a clear:
`1: Tell us your postcode, so we can suggest some places`
kind of indication. This would effectively pre-fill the values in the boxes.

So on the left there would be:

- Near where I live box (location)
- Route (location, finish); people might want to add more than one
- Ward (location); nearby wards are probably also important to be possible
- City (location)
- Groups matching my area
- Draw a map area (replacing the current button)
- Draw a route (replacing the current button)
- Add a point location (replacing the current button)
- (More could be added in future.)

A problem with this is UI concept that it's potentially difficult for a new member (who is naturally unfamiliar with the site) to predict what the effect of these in terms of subscriptions might be. (A user adding a whole city for instance should probably be aware that this potentially brings them a lot of traffic.)

So I think the top-right should show a count of the total number of issues this results in a match for. Although we could do this by discussion thread instead, I think most new users would understand the concept of an issue more easily.

However, this does bring up the thorny issue of whether it makes sense to have multiple locations at the level of `/settings/locations`. Currently a user adds one geometry, and that is a database entry. The page then combines these. The new interface above it seems should just work on the basis of geometries that a user can "add in to the map >>" or "x" delete an entry.

Ideally an entry on the map should have a popup which shows what it is. E.g. adding 'Near where I live box' would show a popup with that or similar text.

This is not 100% thought-through, but hopefully a useful starting point.

**Comment by mvl22** [comment]
The other kind of UI I've been pondering is a more role-based one, i.e.

* Tell us where you live:
* What level of activity do you want to start with:

where the latter is some kind of sliding scale between Newbie and Very active campaigner (lots of messages!), with the setting essentially adding more stuff to the map.

But I think this would be quite hard to make work for different areas in practice. For instance, a small town of 20k people size would probably have the entire area as a sensible default for Newbie, but in London that concept would completely fail.

However, there would still need to be an interface for people to amend/add/remove areas, so I think the first main suggestion above we'd probably need to have in place anyway.

**Comment by nikolai-b** [comment]
I've added the number of issues over the last 3 months.  As the staging DB is a bit old then it might not work there.

Is this enough of an improvement that it can go live?

> Each box would have some kind of tick icon or arrow icon that essentially 'sends' the type to the map. Or each could just have a checkbox to enable/disable each item.

Sorry, I don't understand how this is different from the checkbox there is now that adds that geometry to the map (and when un-checked removes it from the map).  Could you sketch something out (maybe not it words but a very bad mock up)?  Or maybe we should have a call...

**Comment by jf1** [comment]
There's a typo in "estiamte of email traffic" -> estimate...

I can see what you've done to try to make this look as simple as possible for new users however it's not for me and I'm not convinced it's good for the majority.

The fact you can only choose a political area for your 'start/home location' is one limitation. I suppose users could change their postcode to choose another ward/constituency of interest, but this seems much more complex than selecting one from a drop-down or selecting one or more that have been displayed on a map. As an example there's lots of development near many workplaces yet there's almost none near my home.

Group administrators are likely to want to be able to select a Local Authority or a Combined Authority so I hope these geographies will also be available in future.

I don't see the advantage of choosing the local group's area - surely I'd just opt-in to their area by opting-in to the group?
The group is only offered if there's a compatible postcode in the 'home/start location' - c/should the 'destination' postcode be considered too? (eg. Users who live outside of the area a group covers, but regularly commute into the group's area)

Minor point - clearing the 'home/start location' box doesn't affect anything, ie the ward/constituency/group areas can still be selected.

All locations I created were called "a location where you've expressed interest" and there's no option to customise this - such an option would be very useful if users will be able to keep multiple locations (as I hope) rather than a single consolidated one. Ideally the political areas would get their names from the ONS data, perhaps prefixed by the area type (eg WMC for Westminster Constituency), then the user should still be able to customise this.

And I wasn't able to get the 'Route (to destination)' checkbox active. No worries if this is work-in-progress, however I think the misleading word 'Route' under the 'Area Route Point' buttons should be amended asap - I've always thought it would be better described as 'line' or 'line(s)'

**Comment by mvl22** [comment]
Thanks for these thoughts, @jf1. I assume you somehow managed to find the staging server (by changing your /etc/hosts file).

Nikolai and I have discussed the next iteration of the work by phone, and I think this should deal with most of the points you raised. Rather than list the changes here, Nikolai will have something to show on this soon, as it's quite a change from the current prototype, which has the various issues you noticed.

**Comment by mvl22** [comment]
Nikolai,

I see the ward/constituencies/groups functionality is all there now, and that it reacts to map movements.

I think all that remains now is to lay the controls out in a more logically-grouped way.

I've created the following suggested mockup diagram. The concept is:

* Map taking up full width of the page and a good height
* Partly-translucent overlay bar on the left, matching the full height of the map
* Boxes in the overlay, containing each section
* Ticked checkboxes remain when the user has selected an item
* Checkbox lists expand/contract as the map is moved (as you currently have, but vertically rather than horizontally)
* Controls relating to Near me and Route are in the box.
* The inserted polygons can be deleted by selecting the polygon then clicking the delete link in it.
* Top-right has count of issues
* Middle-top of map has a "Move map to..." pan control.

Click on the image to see it at full resolution.

![screen shot 2017-02-19 at 21 28 45](https://cloud.githubusercontent.com/assets/361423/23106783/8d087e22-f6ea-11e6-959c-802b680b2059.png)

**Comment by nikolai-b** [comment]
Thanks for the clear mockup, I will try to do this above Quick Wins.  Some of these are already done and I will keep this up to date with my progress:

- [ ] Map taking up full width of the page and a good height
- [ ] Partly-translucent overlay bar on the left, matching the full height of the map
- [ ] Boxes in the overlay, containing each section
- [x] Ticked checkboxes remain when the user has selected an item
- [ ] Checkbox lists expand/contract as the map is moved (as you currently have, but vertically rather than horizontally)
- [ ] Controls relating to Near me and Route are in the box.
- [ ] The inserted polygons can be deleted by selecting the polygon then clicking the delete link in it.
- [ ] Top-right has count of issues. Doesn't need to mention concept of e-mail traffic.
- [ ] Middle-top of map has a "Move map to..." pan control.
- [x] Only clicking Save saves things. Needs a navigate away warning.
- [ ] This is the complete map for all areas, replacing the current main screen and per-area screens.

**Comment by nikolai-b** [comment]
@mvl22 I've done some more work on this.  Making it take the full width is hard because we have a fixed width as part of our style and I believe that would mean moving to fluid width, I've made a start but can we put that off please?  
Also can we do the popup delete link later?

**Comment by mvl22** [comment]
Surely it can easily be the full width of the current area? I.e. with the whitespace either side, in line with the start and end area of the green menu.

**Comment by nikolai-b** [comment]
Yes that I've done but I thought you wanted full width meaning full width of the screen.
Have you had a chance to review the changes?

**Comment by mvl22** [comment]
Thanks; it's definitely getting there, but not quite fluid or visually polished enough yet.

1) When I load the page without any further interaction, a tooltip is attached to the cursor constantly, referring to "Click cancel to undo changes" (even though I've not made any yet). I presume it is there in order that an existing geometry can be amended, but it seems rather odd at present. I'm not sure how best to handle this. Certainly the cancel stuff is confusing and not relevant. Maybe a geometry should only be editable after clicking on it first or something?

2) The move map stuff should be considered a separate UI device than the area selection. See my design concept above - see the box with "Move map to" in it. At present the current layout is confusing in usability terms because the top item it mixed in with things which manipulate the data.

3) The map is draggable through the side panel on the left, which is not desirable. Possibly the z-index or something is wrong.

4) The titles `Constituencies`, `Wards`, and `Groups` should be a line on their own. It is currently confusing that the first item in each generated checkbox list is essentially inline whereas all the others are block items. The titles should be styled slightly - e.g. bold would be enough to make them a clear title.

5) Sidebar needs some spacing in it, so that the boxes contained in it have space round them - see the example above.

6) Sidebar boxes have black edges. I would always suggest gray for something like this - black is a harsh, unsubtle colour.

7) Any of the three drawing boxes do not change state when they are selected. In other words, it is not clear that they are activated.

8) Save button should be properly aligned with the map, not floating in the middle.

9) The sidebar and the boxes in the sidebar should have a different background colour - again see my example. This helps the boxes stand out.

10) `Route to destination` and `Area around me` are not clickable at first, but still appear the same as others. I recommend these be given say 0.5 opacity to give at least some indication that some other control needs to be manipulated first.

11) Route to destination box needs both the start and finish boxes in. I don't think we should assume that the pan map box is their destination.

12) Route to destination also currently requires exactly the following order to be done: start, finish, click checkbox. Finish then start doesn't put the checkbox in a selectable state.

13) Using the Route to box to search for a place causes the whole sidebar to become larger, with the box outlines stretching over the overlay.

14) Area/route/point and Undo & Cancel should be in a box.

15) The Cancel and Undo controls are barely visible.

16) The `Use current location` button causes the screen to jump upwards. Presumably the browser is picking up the # location, so presumably a `preventDefault()` is missing.

Sorry for the amount of detail and iteration here. I knew this page would need quite a lot of work to get right.

Once the above is in place, I'll do some hallway usability testing.

**Comment by nikolai-b** [comment]
1. Maybe but that would be the opposite of how you've asked me to design things in the past having a feature editable only when clicked.  That goes back to the leaflet draw edit style interface, is that what your suggesting?  It would mean there would be far less customisation so I'd be in favour of using their interface if we can.
1. That means there are an awful lot of UIs on the map.  Not hard to do but are you sure about having so many controls?
1. This is the same as on the issue new (or any other draw map) so I assumed it was intended. Would you like all of the drawing maps to not allow the map to be dragged underneath or just on this page?
1. Done.
1. I cannot see how to make these changes and keep the height of the sidebar, you mention that cancel and undo are barely visible and that your would like more spacing but with all these the height becomes too much especially for smaller screens, should we reduce the nos. of wards, constitutions  and groups shown?
1. Done.
1. Done and update other map drawing
1. Done
1. Done
10. Done 
11. See 1.
12. Done
13. Fixed, not sure how I missed that.
14. Done.
15. Done.
16. Done

**Comment by mvl22** [comment]
> That means there are an awful lot of UIs on the map. Not hard to do but are you sure about having so many controls?

The problem currently is that two entirely different things are being mixed into one UI control area, which is a usability no-no. Essentially there is (1) the requirement to move the map and (2) add areas in.

I think this is then resolved:

> I cannot see how to make these changes and keep the height of the sidebar

**Comment by mvl22** [comment]
I've noticed that the lists, e.g. constituencies aren't ordered. E.g. I'm seeing:

`Constituencies:
South East Cambridgeshire
Cambridge
South Cambridgeshire`

**Comment by nikolai-b** [comment]
2. Done

Also I've added padding to the boxes.

17. Ordering done
18. Thanks, fixed delete
19. Great, removed them.
20. Issue count overlaid

**Comment by mvl22** [comment]
Just to flag up on the staging server that `/overview` is now giving an error 500. (I don't know if this is related to the above changes.)

**Comment by mvl22** [comment]
Thanks.

Could you add padding round the boxes? I.e. make the overlay about 20px wider and add 10px padding either side?

Layout problem with the move map box, whose size obviously isn't so constrained now.

![screen shot 2017-03-03 at 21 52 47](https://cloud.githubusercontent.com/assets/361423/23570455/17bdedca-005c-11e7-9488-1fd7b591bb90.png)

**Comment by nikolai-b** [comment]
Thanks @mvl22, fixed the overview, added a margin round the boxes and unconstrained the search box outside the overlay.

**Comment by nikolai-b** [comment]
@mvl22 have you had a chance to review?

**Comment by mvl22** [comment]
Let's take this live.

There are lots of things I'd like to polish, but I think it's better than the current UI, and I'm keen we ensure that staging doesn't become more complex to manage.

I've not managed to test it in IE yet, because I don't have a machine that I can easily change `/etc/hosts` to hand, but there are no JS errors showing in the Chrome console, which is a good sign.

Once it's live, we'll need to prioritise quickly any IE-related issues.

Thanks for all your hard work on this!

**Comment by nikolai-b** [comment]
Hmm, very good point...
![ie](https://cloud.githubusercontent.com/assets/6289830/23830809/4ec6be0e-070b-11e7-8845-04fe925a6ccc.png)

**Comment by nikolai-b** [comment]
This is live now and I think if there are problems they should be opened as a new issue so I'm closing this.

**Comment by mvl22** [comment]
Just to say that this is in general working fine in IE. There are some minor imperfections, and any bugs I come across I'll note as new issues (e.g. as just posted), but it's basically fine.
